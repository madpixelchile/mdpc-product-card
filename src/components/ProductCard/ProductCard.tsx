
import React,{ createContext } from 'react';
import styles from '../../styles/styles.module.css';

import { useProduct } from '../../hooks/useProduct';
import { Product, ProductContextProps, OnChangeArgs, ProductCartInitialValues, ProductCardHandlers } from '../../interfaces/ProductInterfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
    product: Product;
    children: ( args: ProductCardHandlers ) => React.JSX.Element,
    className?: string;
    style?: React.CSSProperties;
    onChangeProduct?: (args:OnChangeArgs)=> void;
    value?: number;
    initialValues?: ProductCartInitialValues;
}

export const ProductCard = ({ children, product, className = '', style, onChangeProduct, value, initialValues }: Props) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({onChangeProduct, product, value, initialValues });

    return (
        <Provider
            value={{
                counter,
                increaseBy,
                product,
                className,
                maxCount
            }}
        >
            <div 
                className={`${styles.productCard} ${className}`} 
                style={style}
            >
                {
                    children({
                        count: counter,
                        isMaxCountReached,
                        product,
                        maxCount,
                        increaseBy,
                        reset,
                    }) 
                }
            </div>
        </Provider>
    )
}
