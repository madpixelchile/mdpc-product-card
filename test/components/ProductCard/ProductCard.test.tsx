
import React from "react";

import renderer, { act } from 'react-test-renderer'
import { ProductCard } from "../../../src/components/ProductCard/ProductCard";
import { product1 } from "../../data/products";

describe('Product Card Component', () => {

    test('El componente debe de mostrar su children correctamente', () => {

        const productCard = renderer.create(
            <ProductCard product={product1}>
                {
                    () => (
                        <>
                            <h1>hola mundo</h1>
                        </>
                    )
                }
            </ProductCard>
        )

        expect(productCard.toJSON()).toMatchSnapshot();

    })


    test('El counter deberÃ­a de incrementar el contador', () => {

        const productCard = renderer.create(
            <ProductCard product={product1}>
                {
                    ({ count, increaseBy }) => (
                        <>
                            <h1>Product Card</h1>
                            <span>{count}</span>
                            <button onClick={() => increaseBy(2)} >Sumar en test</button>
                        </>
                    )
                }
            </ProductCard>
        )

        let tree = productCard.toJSON();
        expect(tree).toMatchSnapshot();

        act(() => {
            (tree as any).children[2].props.onClick();
        });

        tree = productCard.toJSON();

        // console.log((tree as any).children[1].children[0]);

        expect((tree as any).children[1].children[0]).toBe('2')

    })

});
