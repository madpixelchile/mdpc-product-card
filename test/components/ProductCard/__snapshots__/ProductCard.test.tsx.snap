// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card Component El componente debe de mostrar su children correctamente 1`] = `
<div
  className="productCard "
>
  <h1>
    hola mundo
  </h1>
</div>
`;

exports[`Product Card Component El counter deber√≠a de incrementar el contador 1`] = `
<div
  className="productCard "
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    Sumar en test
  </button>
</div>
`;
